lib/core/src/Cardano/Wallet/Primitive/Model.hs

```diff
@@ -503,7 +512,9 @@ prefilterBlock b u0 = runState $ do
 
         return $ if hasKnownOutput && not hasKnownInput then
```

Comment:
A Tx that fails will have no known output nor any known input, thanks to the implementation of `utxoFromTx`. This means it will hit the last `else` in this `if-else`, which is to just use the `prevUTxO`, but we also want to remove collateral from the UTxO...

I need to write a failing test to demonstrate what I mean.

Reply:
I think this will suffice as an implementation:

```
        let hasKnownInput =
                (if failedScriptValidation (tx ^. #isValidScript)
                 then Set.fromList (collateralInputs tx)
                 else Set.fromList (inputs tx)
                ) `Set.intersection` (Set.fromList ownedAndKnownTxIns)
                /= mempty
```

Reply:
Actually not quite sufficient. We can’t presume that we already know about our inputs that could be used for collateral. The transaction may introduce collateral inputs that we own, so we need to expand the right hand side of the Set intersection too.